lib/elixir/lib/macro.ex

```diff
@@ -2280,4 +2280,138 @@ defmodule Macro do
   defp trim_leading_while_valid_identifier(other) do
     other
   end
+
+  @doc """
+  Default backend for `Kernel.dbg/2`.
+
+  This function provides a default backend for `Kernel.dbg/2`. See the
+  `Kernel.dbg/2` documentation for more information.
+
+  This function:
+
+    * prints information about the given `env`
+    * prints information about `code` and its returned value (using `opts` to inspect terms)
+    * returns the value returned by evaluating `code`
+
+  You can call this function directly to build `Kernel.dbg/2` backends that fall back
+  to this function.
+  """
+  @doc since: "1.14.0"
+  @spec dbg(t, t, Macro.Env.t()) :: t
+  def dbg(code, options, %Macro.Env{} = env) do
+    header = dbg_format_header(env)
+
+    quote do
+      to_debug = unquote(dbg_ast_to_debuggable(code))
+      unquote(__MODULE__).__dbg__(unquote(header), to_debug, unquote(options))
+    end
+  end
+
+  # Pipelines.
+  defp dbg_ast_to_debuggable({:|>, _meta, _args} = pipe_ast) do
+    value_var = Macro.unique_var(:value, __MODULE__)
+    values_acc_var = Macro.unique_var(:values, __MODULE__)
+
+    # Using the private version of unpipe/2 here because Macro.unpipe/1 is deprecated.
+    [start_ast | rest_asts] =
+      asts = for {ast, 0} <- :lists.reverse(unpipe(pipe_ast, _acc = [])), do: ast
+
+    rest_asts = Enum.map(rest_asts, &pipe(value_var, &1, 0))
+
+    initial_acc =
+      quote do
+        unquote(value_var) = unquote(start_ast)
+        unquote(values_acc_var) = [unquote(value_var)]
+      end
+
+    values_ast =
+      for step_ast <- rest_asts, reduce: initial_acc do
+        ast_acc ->
+          quote do
+            unquote(ast_acc)
+            unquote(value_var) = unquote(step_ast)
+            unquote(values_acc_var) = [unquote(value_var) | unquote(values_acc_var)]
+          end
+      end
+
+    quote do
+      unquote(values_ast)
+      {:pipe, unquote(escape(asts)), Enum.reverse(unquote(values_acc_var))}
```

Comment:
I am almost 100% you can convert the asts to string here, instead of escaping them.

Reply:
Done ✅